src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/UniversalCryptoOneShot.cs

```diff
@@ -20,31 +20,137 @@ internal static class UniversalCryptoOneShot
+            // bytesWritten when padding needs to be removed. The padding is "removed" simply
+            // by reporting less of the amount written, then zeroing out the extra padding.
+            if (output.Length >= input.Length)
             {
-                if (output.Length >= input.Length)
+                int bytesTransformed = cipher.TransformFinal(input, output);
+                Span<byte> transformBuffer = output.Slice(0, bytesTransformed);
+
+                try
                 {
-                    bytesWritten = cipher.TransformFinal(input, output);
+                    // validates padding
+                    // This intentionally passes in BlockSizeInBytes instead of PaddingSizeInBytes. This is so that
+                    // "extra padded" CFB data can still be decrypted. The .NET Framework always padded CFB8 to the
+                    // block size, not the feedback size. We want the one-shot to be able to continue to decrypt
+                    // those ciphertexts, so for CFB8 we are more lenient on the number of allowed padding bytes.
+                    bytesWritten = SymmetricPadding.GetPaddingLength(transformBuffer, paddingMode, cipher.BlockSizeInBytes);
+
+                    // Zero out the padding so that the buffer does not contain the padding data "after" the bytesWritten.
+                    CryptographicOperations.ZeroMemory(transformBuffer.Slice(bytesWritten));
                     return true;
                 }
+                catch (CryptographicException)
+                {
+                    // The padding is invalid, but don't leave the plaintext in the buffer.
+                    CryptographicOperations.ZeroMemory(transformBuffer);
+                    throw;
+                }
+            }
 
-                // If no padding is going to be removed, we know the buffer is too small and we can bail out.
+            // If no padding is going to removed, then we already know the buffer is too small
+            // since that requires a buffer at-least the size of the ciphertext. Bail out early.
+            // The second condition is where the output length is short by more than a whole block.
+            // All valid padding is at most one complete block. If the difference between the
+            // output and the input is more than a whole block then we know the output is too small.
+            if (!SymmetricPadding.DepaddingRequired(paddingMode) ||
+                input.Length - cipher.BlockSizeInBytes > output.Length)
+            {
                 bytesWritten = 0;
                 return false;
             }
 
+            // At this point the destination might be big enough but we don't know until we've
+            // transformed the last block, and input is within one block of being the right
+            // size.
+            // For sufficiently small ciphertexts, do them on the stack.
+            const int MaxInStackDecryptionBuffer = 128;
```

Comment:
Possibly worth a comment that this needs to be at least twice as big as the largest block size we support?  (AES with padding will hold back decrypted content in TransformBlock in case TransformFinal provided no data and it already received the last block via TransformBlock).  OK, so that really means 2 * blockSize - 1, but still.

Granted, that means 31, and this is 128... but it feels worth "don't shrink me too much" is useful.